### ISSUE TITLE
Better error message if connection to Vault fails

### ISSUE THREAD

**Issue by hickey** [jantman/vault-aws-creds#10]
When I produced the PR earlier today,  I noticed that the code had been updated so I merged in the changes and then put it in place to run throughout my workflow. Now that I am getting started, I am finding that the script has constant exceptions. Apparently it is trying to connect to the AWS APIs even when just the wrapper function is being requested. 

```
0:0 ᐅ $HOME/lib/python/vault-aws-creds.py -w                                                                                                                        [11:37]
Traceback (most recent call last):
  File "/Users/ghickey/lib/python/vault-aws-creds.py", line 721, in <module>
    exporter = VaultAwsCredExporter(args.config, args.region, args.ttl)
  File "/Users/ghickey/lib/python/vault-aws-creds.py", line 158, in __init__
    self._test_vault_creds()
  File "/Users/ghickey/lib/python/vault-aws-creds.py", line 368, in _test_vault_creds
    raw = self._vault_request('GET', '/v1/auth/token/lookup-self')
  File "/Users/ghickey/lib/python/vault-aws-creds.py", line 335, in _vault_request
    conn.request(method, path, body, headers)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/httplib.py", line 1053, in request
    self._send_request(method, url, body, headers)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/httplib.py", line 1093, in _send_request
    self.endheaders(body)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/httplib.py", line 1049, in endheaders
    self._send_output(message_body)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/httplib.py", line 893, in _send_output
    self.send(msg)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/httplib.py", line 855, in send
    self.connect()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/httplib.py", line 832, in connect
    self.timeout, self.source_address)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/socket.py", line 575, in create_connection
    raise err
socket.error: [Errno 61] Connection refused
```

**Comment by jantman** [comment]
Sorry I haven't had a chance to look at the PR yet, I should this evening or tomorrow.

Are you sure that's trying to connect to AWS APIs? The connection refused seems to be coming from Vault itself.

The traceback above shows:

```
  File "/Users/ghickey/lib/python/vault-aws-creds.py", line 368, in _test_vault_creds
    raw = self._vault_request('GET', '/v1/auth/token/lookup-self')
  File "/Users/ghickey/lib/python/vault-aws-creds.py", line 335, in _vault_request
    conn.request(method, path, body, headers)
```

That seems to be coming from [the code that issues a request to Vault](https://github.com/jantman/vault-aws-creds/blob/master/vault-aws-creds.py#L303-L336).

**Comment by hickey** [comment]
ARG. My vault did not come up after a restart...... OK, change this to a feature request to catch the exception and give a reasonable human error message.

**Comment by jantman** [comment]
Will do. I'll have a new version with this and your PR out sometime this weekend.

**Comment by hickey** [comment]
I just assumed that it was the AWS APIs from the little bit of the URL I was seeing. 

I rolled the code back to about the original code I was running and got the same thing. At which point I broke open the code again and realized it was trying to talk to Vault. Duh :-)

**Comment by hickey** [comment]
Any thoughts on why my capabilities are getting `root` instead of `read`? I am still coming up to speed with Vault and recognize that I may still not be doing something correctly.

**Comment by jantman** [comment]
Do you mean "sudo" instead of "root"? "root" doesn't appear to be a valid [capability](https://www.vaultproject.io/docs/concepts/policies.html#capabilities).

Capabilites are whatever was assigned to you via the policy you're getting in Vault.

That being said, if you _do_ mean sudo, there's a bug in my code regarding that. I just opened #11 for it.

**Comment by hickey** [comment]
No, I mean that my roles are being created with a `root` capability. Look at the body responses to the query of the two roles below. 

```
2018-02-23 12:34:46,347 [DEBUG vault-aws-creds.py:334 - root._vault_request() ] http request to 127.0.0.1:8200 - LIST /v1/pipeline/roles (body: None)
2018-02-23 12:34:46,351 [DEBUG vault-aws-creds.py:337 - root._vault_request() ] Response: HTTP 200 OK
2018-02-23 12:34:46,351 [DEBUG vault-aws-creds.py:338 - root._vault_request() ] Headers: [('date', 'Fri, 23 Feb 2018 20:34:46 GMT'), ('content-length', '184'), ('content-type', 'application/json'), ('cache-control', 'no-store')]
2018-02-23 12:34:46,351 [DEBUG vault-aws-creds.py:344 - root._vault_request() ] Body: {"request_id":"bc562a2f-c1d4-f7af-b205-d3071dda7402","lease_id":"","renewable":false,"lease_duration":0,"data":{"keys":["admin","basic"]},"wrap_info":null,"warnings":null,"auth":null}
2018-02-23 12:34:46,354 [DEBUG vault-aws-creds.py:421 - root.list_roles() ] Checking capabilities for: pipeline/sts/admin
2018-02-23 12:34:46,355 [DEBUG vault-aws-creds.py:334 - root._vault_request() ] http request to 127.0.0.1:8200 - POST /v1/sys/capabilities-self (body: {"path": "pipeline/sts/admin"})
2018-02-23 12:34:46,359 [DEBUG vault-aws-creds.py:337 - root._vault_request() ] Response: HTTP 200 OK
2018-02-23 12:34:46,359 [DEBUG vault-aws-creds.py:338 - root._vault_request() ] Headers: [('date', 'Fri, 23 Feb 2018 20:34:46 GMT'), ('content-length', '207'), ('content-type', 'application/json'), ('cache-control', 'no-store')]
2018-02-23 12:34:46,359 [DEBUG vault-aws-creds.py:344 - root._vault_request() ] Body: {"capabilities":["root"],"request_id":"9dd685f5-9698-3f83-d52e-fc4201c0bf98","lease_id":"","renewable":false,"lease_duration":0,"data":{"capabilities":["root"]},"wrap_info":null,"warnings":null,"auth":null}
2018-02-23 12:34:46,360 [DEBUG vault-aws-creds.py:426 - root.list_roles() ] Capabilities: [u'root']
2018-02-23 12:34:46,360 [DEBUG vault-aws-creds.py:421 - root.list_roles() ] Checking capabilities for: pipeline/sts/basic
2018-02-23 12:34:46,360 [DEBUG vault-aws-creds.py:334 - root._vault_request() ] http request to 127.0.0.1:8200 - POST /v1/sys/capabilities-self (body: {"path": "pipeline/sts/basic"})
2018-02-23 12:34:46,373 [DEBUG vault-aws-creds.py:337 - root._vault_request() ] Response: HTTP 200 OK
2018-02-23 12:34:46,373 [DEBUG vault-aws-creds.py:338 - root._vault_request() ] Headers: [('date', 'Fri, 23 Feb 2018 20:34:46 GMT'), ('content-length', '207'), ('content-type', 'application/json'), ('cache-control', 'no-store')]
2018-02-23 12:34:46,373 [DEBUG vault-aws-creds.py:344 - root._vault_request() ] Body: {"capabilities":["root"],"request_id":"839c2e5e-8668-b527-74b0-86d76fb0e019","lease_id":"","renewable":false,"lease_duration":0,"data":{"capabilities":["root"]},"wrap_info":null,"warnings":null,"auth":null}
2018-02-23 12:34:46,373 [DEBUG vault-aws-creds.py:426 - root.list_roles() ] Capabilities: [u'root']
2018-02-23 12:34:46,373 [DEBUG vault-aws-creds.py:429 - root.list_roles() ] Vault Roles for pipeline/ mountpoint: []
```

This is part of the PR that I submitted. If I am doing something wrong it would be good to figure it out and document the proper process rather than change the code.

**Comment by hickey** [comment]
FYI, I am running Vault 0.9.3.

**Comment by jantman** [comment]
@hickey As far as I can tell, if you're getting capabilites of ``["root"]``, that usually means that either you're using the root token itself, or you're using a token that was created directly by the root token and inherited the default "root" policy.

That's certainly a valid use case for people running Vault in development mode (or perhaps with a minimally-configured instance). I'd never tested it because I primarily use this project at work, where we auth with LDAP or GitHub and are assigned sufficiently-locked-down policies (i.e. nobody gets a "root" policy).

**Comment by jantman** [comment]
This has been merged to master, and is present in the [0.2.6](https://github.com/jantman/vault-aws-creds/releases/tag/0.2.6) release tag that I just created.

Thanks so much!
